R/utils.R

```diff
@@ -140,3 +140,101 @@ edit.data.table = function(name, ...) {
+  #' @param colsub takes the same arguments as .SDcols such as
+  #' `patterns()`, `var1:var2`, a variable, or a vector of characters.
+  #' When TRUE, will return seq_len(length(x))
+  #' @param mode takes character vector of length 1. In the short-term
+  #' there will be different modes based on the call. In the long-term, it's just
+  #' to help the end-user what caused the error
+  #' @return Integer vector with a boolean attribute, `"negate"` which is TRUE when
+  #' `colsub[[1L]]` is `-` or `!`.
+  
+  # see @jangorecki #4174 for how to get the vast majority of this into C
+  
+  x_len = length(x)
+  origsub = colsub
+  
+  # FR #4979 - negative numeric and character indices for SDcols
+  # fix for #5190. colsub[[1L]] gave error when it's a symbol.
+  negate_cols = is.call(colsub) && (colsub[[1L]] == "!" || (colsub[[1L]] == "-" && length(colsub) == 2L))
+  if (negate_cols) colsub = colsub[[2L]]
+
+  if (is.call(colsub)){
+    # fix for #1216, make sure the parentheses are peeled from expr of the form (((1:4)))
+    if (colsub[[1L]] == "(") {
+      colsub = as.list(colsub)[[-1L]]
+      while(length(colsub) > 1L && colsub[[1L]] == "(") colsub = colsub[[-1L]]
+      # give users a second chance with negation for (-1)
+      if (length(colsub) == 2L && colsub[[1L]] == "-" && is.numeric(colsub[[2L]])) {
+        negate_cols = TRUE ##what about -(-1)???
+        colsub = colsub[[2L]]
+      }
+    }
+    if (length(colsub) == 3L && colsub[[1L]] == ":") {
+      if (is.name(colsub[[2L]]) && is.name(colsub[[3L]])){
+        # cols is of the format a:c
+        rnge = chmatch(c(as.character(colsub[[2L]]), as.character(colsub[[3L]])), names(x))
+        if (anyNA(rnge)) stop(gettextf("Not all columns were found in the %s column select statement: [%s]", mode, deparse(origsub), domain = "R-data.table"))
+        cols = rnge[1L]:rnge[2L] 
+      } else if (is.character(colsub[[2L]]) || is.character(colsub[[3L]])){
+        stop(gettextf("When selecting a range of columns in the var_1:var_2 format, quotation marks should not be used. Instead of %s try %s:%s", deparse(origsub), deparse(as.name(colsub[[2L]])), deparse(as.name(colsub[[3L]])), domain = "R-data.table"))
+      } else {
+        # cols is of the format 1:3 or would also allow min(var):max(var)
+        cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+      }
+    } else if (length(colsub) > 1 && colsub[[1L]] == "patterns") {
+      # each pattern gives a new filter condition, intersect the end result
+      x_names = names(x)
+      cols = Reduce(intersect, lapply(as.list(colsub)[-1L], function(col) grep(eval(col, parent.frame(2L)), x_names)))
+    } else {
+      cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+    }
+  } else {
+    # if colsub is also a name in the list, we prioritize it over anything in the environment
+    if (!exists(cols <- as.character(colsub), where = x)) 
+      cols = eval(colsub, parent.frame(2L), parent.frame(2L))
```

Comment:
This is where ```dt[, .SD, .SDcols = V1]``` would allow a variable a variable within the ```dt``` to be picked ahead of anything in the global environment. Would it address your concern if there's a new warning to indicate that there's also a variable with the same name in both the ```dt``` and global environment? To me, if there's a variable within the ```dt```, it should be expected behavior that the ```dt``` variable is prioritized.

Reply:
Using unquoted column name in SDcols is not really necessary there. As well it is not really required for consistency as we do not allow list (v1, v2) syntax, as we do in j arg. Only unquoted form of v1:v2 should be there IMO.